QuoteGeneration/quote_wrapper/quote/id_enclave/id_enclave.cpp

```diff
@@ -288,3 +308,201 @@ sgx_status_t ide_get_pce_encrypt_key(
+        uint8_t crypto_suite,
+        uint16_t cert_key_type,
+        uint32_t key_size,
+        uint8_t* p_public_key,
+        uint32_t ppid_size,
+        uint8_t* ppid)
+{
+    ref_rsa_params_t g_rsa_key = { 0 };  // The private key used to encrypt the PPID.  Only used for PPID_CEARTEXT Cert_Data_Type
+    get_pce_encrypt_decrypt_key(p_pce_target_info, p_ide_report, crypto_suite, cert_key_type, key_size, p_public_key, &g_rsa_key);
+
+    void *rsa_key = NULL;
+    unsigned char* dec_dat = NULL;
+    size_t ppid_size = 0;
+    // Decrypt the PPID with the RSA private key generated with the new key and store it in the blob
+    // Create a private key context
+    /// todo: add a check to see if the private key was lost due to enlave unload or power loss.
+    if (SGX_SUCCESS != sgx_create_rsa_priv2_key(REF_RSA_OAEP_3072_MOD_SIZE,
+                                                REF_E_SIZE_IN_BYTES,
+                                                (const unsigned char*)g_rsa_key.e,
+                                                (const unsigned char*)g_rsa_key.p,
+                                                (const unsigned char*)g_rsa_key.q,
+                                                (const unsigned char*)g_rsa_key.dmp1,
+                                                (const unsigned char*)g_rsa_key.dmq1,
+                                                (const unsigned char*)g_rsa_key.iqmp,
+                                                &rsa_key)) {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+    if (SGX_SUCCESS != sgx_rsa_priv_decrypt_sha256(rsa_key,
+                                                   NULL,
+                                                   (&ppid_size),
+                                                   p_encrypted_ppid,
+                                                   REF_RSA_OAEP_3072_MOD_SIZE)) {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+    //if (sizeof(ciphertext_data.ppid) < ppid_size) {
+    //    ret = REFQE3_ERROR_CRYPTO;
+    //    goto ret_point;
+    //}
+    if (!(dec_dat = (unsigned char*)malloc(ppid_size))) {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+    if (SGX_SUCCESS != sgx_rsa_priv_decrypt_sha256(rsa_key,
+                                                   dec_dat,
+                                                   (&ppid_size),
+                                                   p_encrypted_ppid,
+                                                   REF_RSA_OAEP_3072_MOD_SIZE)) {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+    // Copy in the decrypted PPID
+    memcpy(ppid, dec_dat, ppid_size);
+}
+
+sgx_status_t get_pce_encrypt_decrypt_key(
+        const sgx_target_info_t* p_pce_target_info,
+        sgx_report_t* p_ide_report,
+        uint8_t crypto_suite,
+        uint16_t cert_key_type,
+        uint32_t key_size,
+        uint8_t* p_public_key,
+        ref_rsa_params_t* g_rsa_key)
+{
+    sgx_status_t sgx_status = SGX_SUCCESS;
```

Comment:
Copy of https://github.com/raoulstrackx/SGXDataCenterAttestationPrimitives/blob/e0a57b5d25606f24d7d320ac50df6514117f3a9b/QuoteGeneration/quote_wrapper/quote/enclave/quoting_enclave_3.cpp#L828